### PR TITLE
Allow blocking specific firmware releases by checksum

### PR DESCRIPTION
**Pull request by hughsie**
Fixes https://github.com/fwupd/fwupd/issues/2280

### CODE REVIEW

File: libfwupd/fwupd-client.c [hunk 1]
```diff
@@ -2128,6 +2128,98 @@ fwupd_client_set_approved_firmware (FwupdClient *client,
...
 
+/**
+ * fwupd_client_get_blocked_firmware:
+ * @client: A #FwupdClient
+ * @cancellable: the #GCancellable, or %NULL
+ * @error: the #GError, or %NULL
+ *
+ * Gets the list of blocked firmware.
+ *
+ * Returns: (transfer full): list of remotes, or %NULL
```
Comment (superm1) [inline libfwupd/fwupd-client.c:2139]: not remotes

File: src/fu-engine.c [hunk 2]
```diff
@@ -527,6 +528,7 @@ fu_engine_modify_config (FuEngine *self, const gchar *key, const gchar *value, G
 	const gchar *keys[] = {
 		"ArchiveSizeMax",
 		"DisabledDevices",
+		"DisabledFirmware",
```
Comment (superm1) [inline src/fu-engine.c:531]: BlockedFirmware no?

File: src/fu-history.c [hunk 3]
```diff
@@ -325,6 +343,10 @@ fu_history_create_or_migrate (FuHistory *self, guint schema_ver, GError **error)
 		g_debug ("migrating v%u database by altering", schema_ver);
 		if (!fu_history_migrate_database_v4 (self, error))
 			return FALSE;
+	} else if (schema_ver == 5) {
+		g_debug ("migrating v%u database by altering", schema_ver);
+		if (!fu_history_migrate_database_v5 (self, error))
+			return FALSE;
```
Comment (superm1) [inline src/fu-history.c:349]: don't you also need to do this if someone goes from 1->6 and 2->6 and 3->6 etc?

File: src/fu-util.c [hunk 4]
```diff
@@ -2311,6 +2352,153 @@ fu_util_check_polkit_actions (GError **error)
...
+	devices = fwupd_client_get_history (priv->client, NULL, error);
+	if (devices == NULL)
+		return NULL;
+	device = fu_util_prompt_for_device (priv, devices, error);
+	if (device == NULL)
+		return NULL;
+	release = fu_util_prompt_for_release (priv, fwupd_device_get_releases (device), error);
+	if (release == NULL)
+		return NULL;
+	return fu_util_prompt_for_checksum (priv, fwupd_release_get_checksums (release), error);
```
Comment (superm1) [inline src/fu-util.c:2371]: I think it's going to be confusing that multiple checksums are going to be available to block.  Why not just block the largest (and presumably most secure) checksum?

File: src/fu-util.c [hunk 5]
```diff
@@ -2557,6 +2745,24 @@ main (int argc, char *argv[])
...
+		     "[CHECKSUM]",
+		     /* TRANSLATORS: command description */
+		     _("Blocks a specific firmware from being installed."),
+		     fu_util_unblock_firmware);
+	fu_util_cmd_array_add (cmd_array,
+		     "get-blocked-firmware",
+		     NULL,
+		     /* TRANSLATORS: command description */
+		     _("Gets the list of blocked firmware."),
+		     fu_util_get_blocked_firmware);
```
Comment (superm1) [inline src/fu-util.c:2765]: This same set of commands should be added to fu-tool.c too

File: libfwupd/fwupd-client.c (hunk 1 above)
Comment (hughsie) [reply libfwupd/fwupd-client.c:2139]: Fixed.

File: src/fu-engine.c (hunk 2 above)
Comment (hughsie) [reply src/fu-engine.c:531]: Eeek, indeed, good catch. Fixed.

File: src/fu-history.c (hunk 3 above)
Comment (hughsie) [reply src/fu-history.c:349]: Yes, good catch, fixed.

File: src/fu-util.c (hunk 4 above)
Comment (hughsie) [reply src/fu-util.c:2371]: Valid point. We can use the already-exported `fwupd_checksum_get_best()` -- Fixed.

**Comment by hughsie** [comment]
> Missing bash completion for the new commands

Fixed.

File: src/fu-util.c (hunk 5 above)
Comment (hughsie) [reply src/fu-util.c:2765]: I'm not so sure about this; the `approved` stuff isn't in `fwupdtool` either, and I don't think it's particularly useful to allow the admin running as root to say "don't install foo" which then prevents the (same?) admin from installing that firmware update. I don't think fwupdtool has to have the same featureset as fwupdmgras longer term I want to push the power users using fwupdtool to use the daemon even if the daemon auto-quits straight after it's completed.

**Pull request merged**